pkg/datapath/linux/config/config.go

```diff
@@ -874,7 +874,9 @@ func (h *HeaderfileWriter) writeTemplateConfig(fw *bufio.Writer, e datapath.Endp
 		fmt.Fprintf(fw, "#define USE_BPF_PROG_FOR_INGRESS_POLICY 1\n")
 	}
 
-	if option.Config.ForceLocalPolicyEvalAtSource {
+	// When both the host routing and endpoint routes are enabled we need
+	// to check policies at source.
+	if option.Config.ForceLocalPolicyEvalAtSource || (option.Config.EnableEndpointRoutes && !option.Config.EnableHostLegacyRouting) {
 		fmt.Fprintf(fw, "#define FORCE_LOCAL_POLICY_EVAL_AT_SOURCE 1\n")
```

Comment:
Setting this in the global datapath configuration will force policy evaluation at the source for lxc -> lxc, but also for hostns -> lxc, eth0 -> lxc, and overlay -> lxc because all of those paths call `ipv{4,6}_local_delivery`. Is that really what we want?

AFAICS, that will cause us again to enforce policies twice for all of those paths, when endpoint routes and BPF Host Routing are enabled.

Reply:
Thanks for the comments @pchaigno!

The idea is that `!option.Config.EnableHostLegacyRouting` means that after checking policy at source we will forward the packet directly to the pod using `redirect_peer`, thus bypassing the outer part of the veth. The only case when we trigger the policy program at the outer veth is when traffic gets there directly from the hostns (as endpoint routes are set). Does this make sense for you?

Reply:
Ok, I think I see what I was missing. The packet will then follow the path: bpf_lxc:handle_policy() -> tail_ipv4_ct_ingress_policy_only() -> tail_ipv4_policy() -> ipv4_policy() -> common.h:redirect_ep() -> ctx_redirect_peer(). And for hostns -> lxc, it won't skip the destination's ingress BPF program because of the `needs_backlog` variable, is that right?

Overall, it feels super fragile. The path is complex, with many conditions that could send us into a double enforcement or policy bypass if they ever change the wrong way.
What kind of test coverage do we have for this? Will we notice if `FORCE_LOCAL_POLICY_EVAL_AT_SOURCE` disappears tomorrow (consequence: policy bypass), or if `from_host` is not correctly propagated into `needs_backlog` anymore (consequence: policy bypass), or if [this condition](https://github.com/cilium/cilium/blob/4310b5e5c4a15049d963cbad8791226567743e9c/bpf/bpf_lxc.c#L1881) is extended (consequence: double enforcement)?

Reply:
Hi @pchaigno! We will notice if policy is not applied correctly via the connectivity test. However, we currently don't automatically test dual policy enforcement.

Reply:
That will only be the case if this configuration is covered in some test somewhere. Is it? Do the Datapath Conformance tests enable BPF Host Routing?

Reply:
The host routing is enabled by default unless it is disabled on kernels which support it (= `EnableHostLegacyRouting` is disabled by default.) The net-next, at least, runs with the host routing enabled.

Reply:
Discussed offline. Today the net-next tests automatically disable BPF Host Routing because it is not supported. Once this PR is merged, this "auto-disablement" won't happen anymore and the tests will start testing BPF Host Routing with endpoint routes without any changes to the tests themselves being needed.

LGTM.